synapse/storage/databases/main/presence.py

```diff
@@ -281,20 +282,25 @@ async def should_user_receive_full_presence_with_token(
             True if the user should have full presence sent to them, False otherwise.
         """
 
-        def _should_user_receive_full_presence_with_token_txn(
-            txn: LoggingTransaction,
-        ) -> bool:
-            sql = """
-                SELECT 1 FROM users_to_send_full_presence_to
-                WHERE user_id = ?
-                AND presence_stream_id >= ?
-            """
-            txn.execute(sql, (user_id, from_token))
-            return bool(txn.fetchone())
+        token = await self._get_when_user_should_receive_full_presence(user_id)
+        if token is None:
+            return False
 
-        return await self.db_pool.runInteraction(
-            "should_user_receive_full_presence_with_token",
-            _should_user_receive_full_presence_with_token_txn,
+        return from_token <= token
+
+    @cached()
+    async def _get_when_user_should_receive_full_presence(
+        self, user_id: str
+    ) -> Optional[int]:
+        """Return the presence stream token, if any, which should trigger the
+        user to receive full presence.
```

Comment:
The thing is, aiui it's not a presence stream token that triggers full presence, but rather a user presenting a token *older* than a particular point.

Generally some more words here would be nice.

```suggestion
        """Get the presence token corresponding to the last full presence update for this user.
       
        If the user presents a sync token with a presence stream token at least as old as the result,
        then we need to send them a full presence update.
        
        If this user has never needed a full presence update, returns `None`.
```